### PR TITLE
replace socket write with send with MSG_NOSIGNAL

### PR DESCRIPTION
**Pull request by mekolat**
As seen in core.cpp tmwa is [already supposed to ignore SIGPIPE](https://github.com/themanaworld/tmwa/blob/master/src/high/core.cpp#L145) but it doesn't do so

http://stackoverflow.com/a/6824284
http://stackoverflow.com/a/1705705

### CODE REVIEW

File: src/io/fd.cpp [hunk 1]
```diff
@@ -75,6 +75,10 @@ namespace io
     {
         return ::write(fd, buf, count);
     }
+    ssize_t FD::send(const void *buf, size_t count)
```
Comment (o11c) [inline src/io/fd.cpp]: This is an abuse of the `io::FD` interface.

Any methods on `io::FD` must correspond *exactly* except for the type of the zeroeth argument. Passing a specific flag should be the responsibility of the caller of `io::FD::send`.

**Comment by o11c** [comment]
Given the call to `compat_signal`, there is something horribly wrong if `SIGPIPE` is being generated at all.

Have you added any external library calls to tmwa since I stopped maintaining it, or applied stdio redirection?

Investigate *which* write is generating an unsuppressed `SIGPIPE`, and investigate the environment at that time.

I strongly oppose "change stuff at random and see if it works" as a production strategy.

**Comment by mekolat** [comment]
* no we did not add any library
* the write that does the sigpipe is https://github.com/themanaworld/tmwa/blob/master/src/net/socket.cpp#L156
* debug info: 
```
packet flood detected from Dustman [0x89]
packet flood detected from Dustman [0x89]
packet flood detected from Dustman [0x89]
packet flood detected from Dustman [0x89]
packet flood detected from Dustman [0x89]
packet flood detected from Dustman [0x89]
packet flood detected from Dustman [0x89]
Player [Dustman] has logged off your server
Program received signal SIGPIPE, Broken pipe.
0x00007ffff70df620 in write () from /lib/x86_64-linux-gnu/libc.so.6
```
```
(gdb) bt full
```
```
#1  0x0000000000418808 in tmwa::send_from_fifo (s=0x34f4a60) at src/net/socket.cpp:156
     len = <optimized out>
#2  0x0000000000419497 in tmwa::do_sendrecv (next_ms=...) at src/net/socket.cpp:440
     s = 0x34f4a60
     i = {fd = 38, static DEFAULT_MODE = 438}
     any = <optimized out>
     rfd = {fds = {fds_bits = {5858337886208, 0 <repeats 15 times>}}}
     wfd = {fds = {fds_bits = {14838977724064, 0 <repeats 15 times>}}}
     timeout = {tv_sec = 0, tv_usec = 66966}
#3  0x000000000040e7ad in tmwa_main (argc=<optimized out>, argv=<optimized out>) at src/high/core.cpp:170
     now = <optimized out>
     next = <optimized out>
     args = 0x7fffffffe4b0
#4  0x00007ffff7027b45 in __libc_start_main () from /lib/x86_64-linux-gnu/libc.so.6
No symbol table info available.
#5  0x000000000040e385 in _start ()
```

**Comment by o11c** [comment]
But what are the full contents of the associated `siginfo_t`?

Is there any way that the `sigaction` syscall would fail? Has something changed about the underlying libc or kernel in the new server?

**Comment by mekolat** [comment]
@o11c AH so this would be the problem. wushin says the new server uses a very old kernel

**Comment by o11c** [comment]
Even so I would be surprised if any non-buggy kernel had a problem with these functions, and running on a known-buggy kernel is scary.

I really want `print $_siginfo` from that gdb session.

**Comment by wushin** [comment]
@o11c @mekolat Can we let this one rest for a bit to watch what the new kernel does. I really think it was related to the POS 2.6 kernel.

**Comment by mekolat** [comment]
@wushin I applied all the modifications o11c asked so now it's not a bug fix but rather an improvement

**Comment by wushin** [comment]
@mekolat ok.

**Comment by o11c** [comment]
Also prefer to squash all the syscall additions into one patch and left out actually calling the new syscall (if you even decide to use it).

**Pull request merged**